pallets/lbp/src/trade_execution.rs

```diff
@@ -0,0 +1,150 @@
+
+		let now = T::BlockNumberProvider::current_block_number();
+		let (weight_in, weight_out) = Self::get_sorted_weight(assets.asset_in, now, &pool_data)
+			.map_err(|err| ExecutorError::Error(err.into()))?;
+		let asset_in_reserve = T::MultiCurrency::free_balance(assets.asset_in, &pool_id);
+		let asset_out_reserve = T::MultiCurrency::free_balance(assets.asset_out, &pool_id);
+
+		let fee_asset = pool_data.assets.0;
+
+		if fee_asset == assets.asset_in {
+			let amount_out = hydra_dx_math::lbp::calculate_out_given_in(
+				asset_in_reserve,
+				asset_out_reserve,
+				weight_in,
+				weight_out,
+				amount_in,
+			)
+			.map_err(|_| ExecutorError::Error(Error::<T>::Overflow.into()))?;
+
+			Ok(amount_out) //amount with fee applied as the user is responsible to send fee to the fee collector
+		} else {
+			let calculated_out = hydra_dx_math::lbp::calculate_out_given_in(
+				asset_in_reserve,
+				asset_out_reserve,
+				weight_in,
+				weight_out,
+				amount_in,
+			)
+			.map_err(|_| ExecutorError::Error(Error::<T>::Overflow.into()))?;
+
+			let fee = Self::calculate_fees(&pool_data, calculated_out).map_err(ExecutorError::Error)?;
+			let amount_out_without_fee = calculated_out
+				.checked_sub(fee)
+				.ok_or_else(|| ExecutorError::Error(Error::<T>::Overflow.into()))?;
+
+			Ok(amount_out_without_fee) //amount without fee as the pool is responsible to send fee to the fee collector
+		}
+	}
+
+	fn calculate_buy(
+		pool_type: PoolType<AssetId>,
+		asset_in: AssetId,
+		asset_out: AssetId,
+		amount_out: Balance,
+	) -> Result<Balance, ExecutorError<Self::Error>> {
+		if pool_type != PoolType::LBP {
+			return Err(ExecutorError::NotSupported);
+		}
+
+		let assets = AssetPair { asset_in, asset_out };
+		let pool_id = Self::get_pair_id(assets);
+		let pool_data =
+			<PoolData<T>>::try_get(&pool_id).map_err(|_| ExecutorError::Error(Error::<T>::PoolNotFound.into()))?;
+
+		let now = T::BlockNumberProvider::current_block_number();
+		let (weight_in, weight_out) = Self::get_sorted_weight(assets.asset_in, now, &pool_data)
+			.map_err(|err| ExecutorError::Error(err.into()))?;
+		let asset_in_reserve = T::MultiCurrency::free_balance(assets.asset_in, &pool_id);
+		let asset_out_reserve = T::MultiCurrency::free_balance(assets.asset_out, &pool_id);
+
+		let fee_asset = pool_data.assets.0;
+
+		if fee_asset == assets.asset_out {
+			let fee = Self::calculate_fees(&pool_data, amount_out).map_err(ExecutorError::Error)?;
+			let amount_out_plus_fee = amount_out
+				.checked_add(fee)
+				.ok_or_else(|| ExecutorError::Error(Error::<T>::Overflow.into()))?;
+
+			let calculated_in = hydra_dx_math::lbp::calculate_in_given_out(
+				asset_in_reserve,
+				asset_out_reserve,
+				weight_in,
+				weight_out,
+				amount_out_plus_fee,
+			)
+			.map_err(|_| ExecutorError::Error(Error::<T>::Overflow.into()))?;
+
+			Ok(calculated_in) //TODO: Double check with someone if this is correct
```

Comment:
I am not sure about this. Although it seems it works with this according to the integration tests, I expected to remove the fee from this amount, like in `calculate_sell` line 46, as the pool is responsible for sending the fee to the fee-collector

Reply:
checks out, looks the same as lbp impl, pls remove TODO